### PR TITLE
5485 - fix to keep near cache config to be in sync with map container

### LINKED ISSUE #5485: Memory leak in hazelcast cluster caused by race condition on node join
The way how issue is reproducible in _our_ environment
* Start _node1_
* Load _node1_ with large number of http requests (each request will result into hundreds of hazelcast operations)
* Starting _node2_ when load is running
* You'll see map size node _node2_ growing above configured max size. Issue is _not_ reproducible if load is applied after node joined.

I was unable to reproduce issue using unit test, however when debugging it looks like check in 
MaxSizeChecker#isEvictablePerNode doesn't succeed (hz. v3.4.2) and number of items returned is less than number if item returned by map.size.

Issue is reproducible on hz 3.3.1 as well.

Would be awesome to know if this is known issue or/and if any workaround exist.

### PR DESCRIPTION
**Pull request by PetroSemeniuk**
Hi guys,

  This PR addresses next issues:
* Fixes memory leak which happens if near cache invalidation event was received, but map container for certain map wasn't yet configured (ie. was added later when hazelcast instance was already started). In that case near cache max will have size of max integer.
* Fixes problem when map configuration can't be updated at runtime (I believe this functionality is part of Hazecast Management Console). While map container can be successfully reconfigured, corresponding near cache will container stale value of max size.

On my side all tests are passing. Please let me know whether anything else need to be done.

Cheers
Petro

Fixes #5485

### CODE REVIEW

**Comment by PetroSemeniuk** [comment]
@devOpsHazelcast hey, al tests were 100% green on my side.

I don't ave access to your cloudbees setup, can you post it here what failures were?

**Comment by Donnerbart** [comment]
It was not a test failure, but CheckStyle issue. Please use the Checkstyle and Findbugs profile locally to find these :)

```
00:00:29.422 [INFO] --- maven-checkstyle-plugin:2.12:checkstyle (default) @ hazelcast ---
00:00:44.697 [INFO] Starting audit...
00:00:44.697 /scratch/jenkins/workspace/Hazelcast-3.x-pr-builder/hazelcast/src/main/java/com/hazelcast/map/impl/NearCache.java:87:28: '{' is not preceded with whitespace.
00:00:44.697 /scratch/jenkins/workspace/Hazelcast-3.x-pr-builder/hazelcast/src/main/java/com/hazelcast/map/impl/NearCache.java:89:48: ':' is not preceded with whitespace.
```

File: hazelcast/src/main/java/com/hazelcast/map/impl/NearCache.java [hunk 1]
```diff
@@ -86,12 +81,26 @@ public NearCache(String mapName, NodeEngine nodeEngine) {
         nearCacheStats = new NearCacheStatsImpl();
         lastCleanup = Clock.currentTimeMillis();
         serializationService = nodeEngine.getSerializationService();
+        nearCacheSizeEstimator = mapContainer.getNearCacheSizeEstimator();
+    }
+
+    public int getMaxSize(){
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/map/impl/NearCache.java]: Missing whitespace `){`

File: hazelcast/src/main/java/com/hazelcast/map/impl/NearCache.java [hunk 2]
```diff
@@ -86,12 +81,26 @@ public NearCache(String mapName, NodeEngine nodeEngine) {
...
+        final int maxSize = mapContainer.getMapConfig().getNearCacheConfig().getMaxSize();
+        return maxSize <= 0 ? Integer.MAX_VALUE: maxSize;
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/map/impl/NearCache.java]: Here is the other one.

File: hazelcast/src/main/java/com/hazelcast/map/impl/NearCache.java (hunk 1 above)
Comment (PetroSemeniuk) [inline hazelcast/src/main/java/com/hazelcast/map/impl/NearCache.java]: Fixed. Thanks!

File: hazelcast/src/main/java/com/hazelcast/map/impl/NearCache.java (hunk 2 above)
Comment (PetroSemeniuk) [inline hazelcast/src/main/java/com/hazelcast/map/impl/NearCache.java]: Fixed. Thanks!

**Comment by ahmetmircik** [comment]
Hi @PetroSemeniuk, can you also provide separate tests for the fixed issues?

**Comment by gurbuzali** [comment]
https://github.com/hazelcast/hazelcast/issues/5485

**Comment by PetroSemeniuk** [comment]
@ahmetmircik I added test for max size check. Would that be enough?

**Comment by ahmetmircik** [comment]
Hi @PetroSemeniuk, can you please also write a test for the other case? Btw, we are not fully  supporting runtime configuration for `IMap` atm only some sub-set of  `IMap` config can be updateable over management center.

**Comment by PetroSemeniuk** [comment]
Hi @ahmetmircik , I added unit test for another use case as well. 

I'm aware that runtime configuration of IMap officially is not fully supported (few related links: http://stackoverflow.com/questions/21379171/hazelcast-mapstoreconfig-ignored, https://github.com/hazelcast/hazelcast/issues/592 and https://github.com/hazelcast/hazelcast/issues/380) however in our case almost all of our imaps are configured dynamically (we have more than hundred of imaps, holding tens of thousand of items and getting zillions of hits. For us updating max size is the most crucial (I believe there should be at least few more customers with this need) and this PR addresses this.

It doesn't compromise near cache performance or architecture in any way and simply makes reasoning about config state of near cache more straightforward.

File: hazelcast/src/test/java/com/hazelcast/map/EvictionTest.java [hunk 3]
```diff
@@ -107,6 +116,62 @@ public void testGetEntryView_withTTL() throws Exception {
...
+                .setNearCacheConfig(nearCacheConfig);
+
+        final TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(1);
+        final HazelcastInstance h = factory.newHazelcastInstance(cfg);
+        final IMap<Object, Object> map = h.getMap(mapName);
+        for (Integer i = 0; i < size * 2; i++) {
+            map.put(i.toString(), i.toString());
+            map.get(i.toString());//trigger put into near cache
+        }
+        sleepSeconds(3); //wait for eviction
```
Comment (ahmetmircik) [inline hazelcast/src/test/java/com/hazelcast/map/EvictionTest.java]: Can you change `sleepSeconds` with `assertTrueEventually `? 

```
assertTrueEventually(new AssertTask() {
                @Override
                public void run() throws Exception {
                     NearCacheStats originalNearCacheStats = map.getLocalMapStats().getNearCacheStats();
                     assertThat(originalNearCacheStats.getOwnedEntryCount(), lessThan((long)size));
                     assertThat(originalNearCacheStats.getOwnedEntryCount(), greaterThan((long) (size/2)));                
              }
            });

```

File: hazelcast/src/test/java/com/hazelcast/map/EvictionTest.java [hunk 4]
```diff
@@ -107,6 +116,62 @@ public void testGetEntryView_withTTL() throws Exception {
...
+        final MapConfig halvedMapConfig = mapConfig
+                .setMaxSizeConfig(halvedMaxSizeConfig)
+                .setNearCacheConfig(halvedNearCacheConfig);
+        mapContainer.setMapConfig(halvedMapConfig);
+
+        for (Integer i = 0; i < size * 2; i++) {
+            map.put(i.toString(), i.toString());
+            map.get(i.toString());//trigger put into near cache
+        }
+        sleepSeconds(3); //wait for eviction
```
Comment (ahmetmircik) [inline hazelcast/src/test/java/com/hazelcast/map/EvictionTest.java]: same here.

File: hazelcast/src/test/java/com/hazelcast/map/EvictionTest.java (hunk 3 above)
Comment (PetroSemeniuk) [inline hazelcast/src/test/java/com/hazelcast/map/EvictionTest.java]: Done.

File: hazelcast/src/test/java/com/hazelcast/map/EvictionTest.java (hunk 4 above)
Comment (PetroSemeniuk) [inline hazelcast/src/test/java/com/hazelcast/map/EvictionTest.java]: Done.

**Comment by PetroSemeniuk** [comment]
@ahmetmircik all done. Thanks for looking into that!

P.S. Sorry for late reply, was busy optimising memory in our own product and missed your message.

**Comment by ahmetmircik** [comment]
Thanks for this PR but since we don't support dynamic configuration, i tend to close it and will revisit it later if we decide to have dynamic configuration.

**Pull request closed without merging**